syntax = "proto3";

service MusicalServer {

    rpc GenerateMusicalFormParameters (GenerateMusicalFormParametersRequest) returns (GenerateMusicalFormParametersResponse);

}

message Measure {
    int32 start_beat = 1;
    int32 end_beat = 2;
}

message TimeSignature {
    int32 num_beats_per_measure = 1;
    int32 beat_note_value = 2;
}

// Pass in either a numerical value, or a TempoMarking
message Tempo {
    int32 beats_per_min = 1;
    optional TempoMarking marking = 2;
}

message Motif {
    repeated int32 notes = 1;
}

message Passage {
    repeated Motif motives = 1;
}

message MusicalParameters {
    TimeSignature time_signature = 1;
    Tempo tempo = 2;
}


// TODO: first endpoint for server, request parameters
enum MusicalForm {
    SCHERZO = 0;
    WALTZ = 1;
    MINUET = 2;
}

enum TempoMarking {
    GRAVE = 0; // 20 - 40 bpm
    LENTO = 1; // 40 - 45 bpm
    LARGO = 2; // 45 - 50 bpm
    ADAGIO = 3;
    ADAGIETTO = 4;
    ANDANTE = 5;
    MODERATO = 6;
    ALLEGRETTO = 7;
    ALLEGRO = 8;
    VIVACE = 9;
    PRESTO = 10;
    PRESTISSIMO = 11;
}

message GenerateMusicalFormParametersRequest {
    MusicalForm form = 1;
}

message GenerateMusicalFormParametersResponse {
    MusicalParameters parameters = 1;
}

